app/src/main/java/com/nexenio/bleindoorpositioningdemo/ui/beaconview/chart/BeaconLineChart.java

```diff
@@ -317,72 +323,78 @@ protected void drawBeacons(Canvas canvas) {
-                if (advertisingPacket.getTimestamp() < lastAdvertisingPacket.getTimestamp() + 5000) {
-                    canvas.drawLine(
-                            lastLinePoint.x,
-                            lastLinePoint.y,
-                            currentLinePoint.x,
-                            currentLinePoint.y,
-                            linePaint
-                    );
-                }
-            } else {
-                lastLinePoint = new PointF();
-            }
+            drawNextPoint(canvas, beacon, advertisingPacket);
+        }
+        currentBeaconIndex++;
+        fadeLastPoint(canvas);
+    }
 
-            lastLinePoint.x = currentLinePoint.x;
-            lastLinePoint.y = currentLinePoint.y;
-            lastAdvertisingPacket = advertisingPacket;
-            currentBeaconIndex++;
+    protected void prepareDraw(Beacon beacon) {
+        lastLinePoint = null;
+        lastAdvertisingPacket = null;
+
+        linePaint.setShader(createLineShader(getBeaconColor(beacon, coloringMode, currentBeaconIndex)));
+        linePaint.setAlpha(128);
+    }
+
+    protected void drawNextPoint(Canvas canvas, Beacon beacon, AdvertisingPacket advertisingPacket) {
+        canvas.drawCircle(currentLinePoint.x, currentLinePoint.y, pixelsPerDip * 1.5f, linePaint);
+
+        if (lastLinePoint != null && lastAdvertisingPacket != null) {
+            if (advertisingPacket.getTimestamp() < lastAdvertisingPacket.getTimestamp() + WINDOW_5_SECONDS) {
+                canvas.drawLine(
+                        lastLinePoint.x,
+                        lastLinePoint.y,
+                        currentLinePoint.x,
+                        currentLinePoint.y,
+                        linePaint
+                );
+            }
+        } else {
+            lastLinePoint = new PointF();
         }
 
+        lastLinePoint.x = currentLinePoint.x;
+        lastLinePoint.y = currentLinePoint.y;
+        lastAdvertisingPacket = advertisingPacket;
+    }
+
+    protected void fadeLastPoint(Canvas canvas) {
         if (lastLinePoint != null) {
             linePaint.setAlpha(255);
             canvas.drawCircle(lastLinePoint.x, lastLinePoint.y, pixelsPerDip * 4, linePaint);
         }
     }
 
-    protected float getValue(Beacon beacon, AdvertisingPacket advertisingPacket, long windowLength) {
-        List<AdvertisingPacket> recentAdvertisingPackets = null;
-        int[] recentRssis;
-        float meanRssi;
+    protected float getValue(Beacon beacon, AdvertisingPacket advertisingPacket) {
+        List<AdvertisingPacket> recentAdvertisingPackets = new ArrayList<>();
+        int rssi;
 
         // make sure that the window size is at least 10 seconds when we're looking for the frequency
-        windowLength = (valueType != VALUE_TYPE_FREQUENCY) ? windowLength : Math.max(windowLength, 10000);
+        long windowLength = (valueType != VALUE_TYPE_FREQUENCY) ? this.windowLength : Math.max(this.windowLength, WINDOW_10_SECONDS);
```

Comment:
This should be done while initialization and not again and again. Also the constant in this case should be named something like `MINIMUM_WINDOW_LENGTH` or use the value `TimeUnit.SECONDS.toMillis(10)` directly.

Reply:
This can't be done during initialization, as `this.windowLength` and `valueType` may change at runtime.